dbms/src/TestUtils/FunctionTestUtils.cpp

```diff
@@ -115,39 +115,48 @@ void blockEqual(
     }
 }
 
+std::pair<ExpressionActionsPtr, String> buildFunction(Context & context, const String & func_name, const ColumnsWithTypeAndName & columns, const ColumnNumbers & argument_column_numbers, const TiDB::TiDBCollatorPtr & collator)
+{
+    std::vector<NameAndTypePair> source_columns;
+    for (const auto & column : columns)
+        source_columns.emplace_back(column.name, column.type);
+    DAGExpressionAnalyzer analyzer(source_columns, context);
+    ExpressionActionsChain chain;
+    auto last_step = analyzer.initAndGetLastStep(chain);
+    auto result_name = DB::FunctionBuildHelper::buildFunction(ColumnsFunctionExpr(columns, argument_column_numbers, func_name, collator), &analyzer, last_step.actions);
+    return std::make_pair(last_step.actions, result_name);
+}
+
+ColumnsWithTypeAndName toColumnsWithUniqueName(const ColumnsWithTypeAndName & columns)
+{
+    ColumnsWithTypeAndName columns_with_distinct_name = columns;
+    std::string base_name = "col";
+    for (size_t i = 0; i < columns.size(); i++)
+    {
+        columns_with_distinct_name[i].name = base_name + std::to_string(i);
+    }
+    return columns_with_distinct_name;
+}
 
 ColumnWithTypeAndName executeFunction(Context & context, const String & func_name, const ColumnsWithTypeAndName & columns, const TiDB::TiDBCollatorPtr & collator)
 {
-    auto & factory = FunctionFactory::instance();
-
-    Block block(columns);
-    ColumnNumbers cns;
-    for (size_t i = 0; i < columns.size(); ++i)
-        cns.push_back(i);
-
-    auto bp = factory.tryGet(func_name, context);
-    if (!bp)
-        throw TiFlashTestException(fmt::format("Function {} not found!", func_name));
-    auto func = bp->build(columns, collator);
-    block.insert({nullptr, func->getReturnType(), "res"});
-    func->execute(block, cns, columns.size());
-    return block.getByPosition(columns.size());
+    ColumnNumbers argument_column_numbers;
+    auto columns_with_distinct_name = toColumnsWithUniqueName(columns);
+    for (size_t i = 0; i < columns_with_distinct_name.size(); i++)
+        argument_column_numbers.push_back(i);
+    auto [actions, result_name] = buildFunction(context, func_name, columns_with_distinct_name, argument_column_numbers, collator);
+    Block block(columns_with_distinct_name);
+    actions->execute(block);
+    return block.getByName(result_name);
```

Comment:
maybe extract the code into a function is better since line156-159 have the same logic.

Reply:
Done